File: LC120Triangle.java
package leetcode;

import java.util.ArrayList;
import java.util.List;

/**
 * https://leetcode.com/problems/triangle/
 */
public class LC120Triangle {
    public int minimumTotal(List<List<Integer>> triangle) {
        if (triangle == null || triangle.size() == 0 || triangle.get(0).size() == 0) {
            return 0;
        }
        int height = triangle.size();
        int[] sums = new int[triangle.get(height - 1).size()];
        sums[0] = triangle.get(0).get(0);
        for (int l = 1; l < height; l++) {
            List<Integer> line = triangle.get(l);
            int n = line.size();
            for (int i = n - 1; i >= 0; i--) {
                int left = i == 0 ? Integer.MAX_VALUE : sums[i - 1];
                int right = i == n - 1 ? Integer.MAX_VALUE : sums[i];
                sums[i] = Math.min(left, right) + line.get(i);
            }
        }
        int min = sums[0];
        for (int i = 1; i < sums.length; i++) {
            if (sums[i] < min) {
                min = sums[i];
            }
        }
        return min;
    }

    private List<List<Integer>> construct(int[][] compact) {
        List<List<Integer>> triangle = new ArrayList<>();
        for (int l = 0; l < compact.length; l++) {
            List<Integer> line = new ArrayList<>();
            triangle.add(line);
            for (int i = 0; i < compact[l].length; i++) {
                line.add(compact[l][i]);
            }
        }
        return triangle;
    }

    public static void main(String[] arguments) {
        LC120Triangle solution = new LC120Triangle();
        int[][] triangle = {
                {2},
                {3, 4},
                {6, 5, 7},
                {4, 1, 8, 3}
        };
        System.out.println(solution.minimumTotal(solution.construct(triangle)));
    }
}

File: algorithms/code/leetcode/lc120_triangle/lc120_triangle.rs
// https://leetcode.com/problems/triangle/
// #medium

pub struct Solution;

impl Solution {
    pub fn minimum_total(triangle: Vec<Vec<i32>>) -> i32 {
        if triangle.is_empty() || triangle[0].is_empty() {
            return 0;
        }
        let height = triangle.len();
        let mut sums: Vec<i32> = vec![0; triangle[height - 1].len()];
        sums[0] = triangle[0][0];
        for line in triangle.iter().take(height).skip(1) {
            let n = line.len();
            for i in (0..n).rev() {
                let left = if i == 0 { i32::MAX } else { sums[i - 1] };
                let right = if i == n - 1 { i32::MAX } else { sums[i] };
                sums[i] = std::cmp::min(left, right) + line[i];
            }
        }
        *sums.iter().min().unwrap()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_example() {
        let triangle = vec![vec![2], vec![3, 4], vec![6, 5, 7], vec![4, 1, 8, 3]];
        assert_eq!(Solution::minimum_total(triangle), 11);
    }

    #[test]
    fn test_nothing() {
        assert_eq!(Solution::minimum_total(vec![]), 0);
        assert_eq!(Solution::minimum_total(vec![vec![]]), 0);
    }
}
